### ISSUE TITLE
Unwanted vertices when tracing

### ISSUE THREAD

**Issue by aharfoot** [qgis/QGIS#52935]
### What is the bug or the crash?

When tracing whilst digitising, vertices will be added to the traced feature where intersections with other features occur, despite no vertex existing on the feature being traced.

This is not desirable, particularly in situations where the absense of a common vertex implies a lack of network connectivity.

If the tracing does not follow intersecting features, then additional vertices should not be added.

### Steps to reproduce the issue

[Polylines_TracingBug.zip](https://github.com/qgis/QGIS/files/11355417/Polylines_TracingBug.zip)

1. Load the project embedded in the attached zipped GPKG
2. The red and green polylines are on the same layer (Polylines) and intersect either at vertices or not.
3. Digitise a new polyline following the green line into the TraceTest layer using tracing mode:
![image](https://user-images.githubusercontent.com/3789465/235201097-c414cb4e-5821-4169-bc9b-49601beceeb2.png)

The newly digitised polyline has vertices at locations where other lines intersect it, but where there isn't a vertex on the green line that was traced.

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.28.5-Firenze | QGIS code revision | 50adba36f2
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.9.5
GDAL/OGR version | 3.6.3
PROJ version | 9.2.0
EPSG Registry database version | v10.082 (2023-02-06)
GEOS version | 3.11.2-CAPI-1.17.2
SQLite version | 3.41.1
PDAL version | 2.5.2
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.6
QScintilla2 version | 2.13.1
OS version | Windows 10 Version 2009
  |   |   |  
Active Python plugins
GroupStats | 2.2.7
processing_r | 3.1.1
valuetool | 3.0.15
db_manager | 0.1.20
processing | 2.12.99

</body></html>QGIS version
3.28.5-Firenze
QGIS code revision
[50adba36f2](https://github.com/qgis/QGIS/commit/50adba36f2)
Qt version
5.15.3
Python version
3.9.5
GDAL/OGR version
3.6.3
PROJ version
9.2.0
EPSG Registry database version
v10.082 (2023-02-06)
GEOS version
3.11.2-CAPI-1.17.2
SQLite version
3.41.1
PDAL version
2.5.2
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009

Active Python plugins
GroupStats
2.2.7
processing_r
3.1.1
valuetool
3.0.15
db_manager
0.1.20
processing
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

**Comment by troopa81** [comment]
Not sure I understand correctly what you described, but I failed to reproduce the issue in master

![image](https://github.com/qgis/QGIS/assets/14358135/bc0b870a-bdab-4cbd-b9b1-5c3d2bf8ea74)

I have no intersection vertex at all instead when an other vertex was already there. Could you make a video of your issue so I can understand exactly what the issue is (or maybe it's already fixed)

**Comment by aharfoot** [comment]
Hi Julien,

Having reread my instructions, I realise that I said follow the green line, when I should have said the red one, however, provided that the digitising in your screenshot was done with Tracing mode enabled then you are demonstrating the same effect (compare with my screenshot below of the corresponding outcome in 3.28.6). It would appear that this is fixed in Master. I will check and confirm.

![image](https://github.com/qgis/QGIS/assets/3789465/0637ea97-aa6a-4d39-8f59-0f9d581902e8)

**Comment by aharfoot** [comment]
Here's a video of the steps I've taken when reproducing these.
![TracingExample](https://github.com/qgis/QGIS/assets/3789465/5295de42-9191-45ef-8587-15b5d4779cd4)

**Comment by troopa81** [comment]
Sorry I didn't know about tracing mode, so that's why I fail to understand your issue. 

To sum up, in this video, first traced geometry need the intersection vertex, but in the second one, you would like to remove the vertex because it's not relevant.

[tracingMode.webm](https://github.com/qgis/QGIS/assets/14358135/4412a9b6-5cd3-4635-93c3-3425915a5fd8)

I'm wondering if this is really an issue (I have no opinion) because maybe some users find the actual behavior legit. Actually, we insert all geometries involved in a graph and try to find the shortest path (and so the intersection vertices pop up).

@wonder-sk as you were the original developer of this, may I have your opinion?

**Comment by aharfoot** [comment]
I can understand why this is happenning, and agree that the behaviour may be desirable in some circumstances, but is certainly not desirable in others - ironically one of these is tracing on a graph where the presence of a vertex on an intersection implies connectivity, whilst the absence does not, think of an overpass vs crossroads in a road network. The tracing functionality exists to reuse elements of existing features exactly, but this behaviour results in features that no longer align to the source.

One option would be to toggle the insertion of extra vertices (presumably whether the trace graph considers intersections without vertices as nodes or not), linking it to the 'Snapping on Intersection' toggle button on the Snapping toolbar. Indeed, the 'Snapping on Intersection' mode would need to be enabled to replicate this behaviour without using the tracing functionality

**Comment by wonder-sk** [comment]
> as you were the original developer of this, may I have your opinion?

I do not really have a strong opinion on this. I think the current behavior is acceptable. I could imagine having a switch to enable/disable tracing on intersections of features when they do not share a vertex (but I would oppose having it linked with "snapping on intersection" which has different meaning - I think this should have a dedicated toggle button)

**Comment by aharfoot** [comment]
@troopa81 Yes, I was thinking of option 1, making a clear distinction between the two modes - with or without snapping (and by extension tracing) to intersections where there isn't a vertex.

**Comment by aharfoot** [comment]
@wonder-sk , I disagree that this distinction is different to the "snapping on intersection" mode that currently exists - the two are identical in their behaviour in that when the "snapping on intersection" mode is enabled, vertices can be generated in the digitised feature at points of intersection between geometries. The Tracing tool is currently behaving as if the "snapping on intersection" mode is permanently enabled.
To put it a different way, if the tracing tool wasn't available, then to reproduce the features in the videos above, the "snapping on intersection" mode would have to be enabled.

**Comment by uclaros** [comment]
This is valid again, since #62345 disabled the fix.
One can still enable it with the following python snippet if not affected by the crashes:
```python
QgsMapCanvasTracer.tracerForCanvas(iface.mapCanvas()).setAddPointsOnIntersectionsEnabled(False)
```
